sql/pipelines/src/main/scala/org/apache/spark/sql/pipelines/graph/CoreDataflowNodeProcessor.scala

```diff
@@ -0,0 +1,232 @@
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.AnalysisException
+import org.apache.spark.sql.catalyst.TableIdentifier
+import org.apache.spark.sql.pipelines.graph.DataflowGraphTransformer.{
+  TransformNodeFailedException,
+  TransformNodeRetryableException
+}
+
+/**
+ * Core processor that is responsible for analyzing each flow and sort the nodes in
+ * topological order
+ */
+class CoreDataflowNodeProcessor(rawGraph: DataflowGraph) {
+
+  private val flowResolver = new FlowResolver(rawGraph)
+
+  // Map of input identifier to resolved [[Input]].
+  private val resolvedInputs = new ConcurrentHashMap[TableIdentifier, Input]()
+  // Map & queue of resolved flows identifiers
+  // queue is there to track the topological order while map is used to store the id -> flow
+  // mapping
+  private val resolvedFlowNodesMap = new ConcurrentHashMap[TableIdentifier, ResolvedFlow]()
+  private val resolvedFlowNodesQueue = new ConcurrentLinkedQueue[ResolvedFlow]()
+
+  private def processUnresolvedFlow(flow: UnresolvedFlow): ResolvedFlow = {
+    val resolvedFlow = flowResolver.attemptResolveFlow(
+      flow,
+      rawGraph.inputIdentifiers,
+      resolvedInputs.asScala.toMap
+    )
+    resolvedFlowNodesQueue.add(resolvedFlow)
+    resolvedFlowNodesMap.put(flow.identifier, resolvedFlow)
+    resolvedFlow
+  }
+
+  /**
+   * Processes the node of the graph, re-arranging them if they are not topologically sorted.
+   * Takes care of resolving the flows and virtualization if needed for the nodes.
+   * @param node The node to process
+   * @param upstreamNodes Upstream nodes for the node
+   * @return
+   */
+  def processNode(node: GraphElement, upstreamNodes: Seq[GraphElement]): Seq[GraphElement] = {
+    node match {
+      case flow: UnresolvedFlow => Seq(processUnresolvedFlow(flow))
+      case failedFlow: ResolutionFailedFlow => Seq(processUnresolvedFlow(failedFlow.flow))
+      case table: Table =>
+        // Ensure all upstreamNodes for a table are flows
+        val flowsToTable = upstreamNodes.map {
+          case flow: Flow => flow
+          case _ =>
+            throw new IllegalArgumentException(
+              s"Unsupported upstream node type for table ${table.displayName}: " +
+              s"${upstreamNodes.getClass}"
+            )
+        }
+        val resolvedFlowsToTable = flowsToTable.map { flow =>
+          resolvedFlowNodesMap.get(flow.identifier)
+        }
+
+        // Assign isStreamingTable (MV or ST) to the table based on the resolvedFlowsToTable
+        val tableWithType = table.copy(
+          isStreamingTableOpt = Option(resolvedFlowsToTable.exists(f => f.df.isStreaming))
+        )
+
+        // Table will be virtual in either of the following scenarios:
+        // 1. If table is present in context.fullRefreshTables
+        // 2. If table has any virtual inputs (flows or tables)
+        // 3. If the table pre-existing metadata is different from current metadata
```

Comment:
Not sure I follow this comment - it seems like `resolvedInputs` will always contain a pointer to a VirtualTableInput for any table being resolved?

Reply:
Yeah, that's true. I removed this for now.